### DISCUSSION TITLE
How do I extend TranslationReader - or rather how do I tell Symfony to use my extended TranslationReader class?

### DISCUSSION THREAD

**Discussion by php4fan** [symfony/symfony#53957] [Q&A]
I need to **change the way Symfony loads translation files**. (I'm looking at Symfony 5 btw, because for the moment I'm stuck with PHP 7.4. Hopefully this doesn't change too much in Symfony 6 or 7).

Specifically, I need to organize translations of the same language and same domain (the default one, 'messages') into multiple files.
I do _not_ want to define additional domains.

So, for example, my `translations` folder may contain the following files:
```
messages.xyz.en.yaml
messages.abc.en.yaml
messages.xyz.es.yaml
messages.abc.es.yaml
```
where `messages.xyz.*` and `messages.abc.*` contain different, non-overlapping sets of phrases, all of them belonging to the `messages` domain, and I want Symfony to treat them exactly as if they were merged into a single file per domain per language.

I see that the reading of files is done by `Symfony\Component\Translation\Reader\TranslationReader` in the `read()` method, so I could create **a class that extends `TranslationReader` and overrides `read()`**. I know how to do that.

Now, I see that `vendor/symfony/framework-bundle/Resources/config/translation.php` does:
```
$container->services()
        //.....
        ->set('translation.reader', TranslationReader::class)
        //.....
```

So, **if only I could override that with:**
```
$container->services()
        //.....
        ->set('translation.reader', MyOwnExtendedTranslationReader::class)
        //.....
```
I should be done.

How can I do that?

I would be happy with somehow **telling Symfony to use my own modified version of `translation.php`** instead of the default one.

Is there a simple way of doing that?

**Comment by MatTheCat** [comment]
See https://symfony.com/doc/current/bundles/override.html#services-configuration.

Youʼll probably need a compiler pass.

**Comment by php4fan** [comment]
> Youʼll probably need a compiler pass.

Thank you! So I tried adding this in my `Kernel.php`:
```
    public function process(ContainerBuilder $container) {
        $container->getDefinition('translation.reader')->setClass(MultiFileTranslationReader::class);
    }
```
but my class `MultiFileTranslationReader` doesn't seem to be used at all. Its `read()` method never gets called. I verified that by adding a `die` statement at the beginning of it. I even tried adding a `die` statement in its constructor, and it has no effect; so my class never even gets instantiated.

The `Kernel::process()` method does get called (if I put a die statement in it it dies).

Any ideas?

**Reply by MatTheCat** [reply]
Indeed `TranslationReader` is only used by translation related commands; it’s the FrameworkBundle’s `Translator` which will dump the translations’ catalogs. You may try overriding its `addResource` method, but I’m not sure it will be enough (nor easy).
